### PR TITLE
autoPatchelfHook: Add dependencies for '.note.dlopen' to rpath

### PR DESCRIPTION
**Pull request by ElvishJerricco**
## Description of changes

This allows `autoPatchelfHook` to resolve `.note.dlopen` dependencies. The main motivation for this is #307068, where we intend to use this hook to automatically add the `.note.dlopen` dependencies of systemd 256 to its rpath. This is a little strange, as systemd is built from source and we haven't historically had to modify its rpath, but this seemed like the most obvious solution.

This also *slightly* changes the handling of libc. Instead of *always* ignoring libc dependencies, this attempts to retain libc in binaries' rpath if they had it originally.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
@layus ping

File: pkgs/build-support/setup-hooks/auto-patchelf.py [hunk 1]
```diff
@@ -38,21 +40,42 @@ def is_dynamic_executable(elf: ELFFile) -> bool:
...
+    Extracts dependencies from the .note.dlopen section, which is used by
+    systemd to store the list of libraries that are opened with dlopen.
+    https://systemd.io/ELF_DLOPEN_METADATA/
+    """
+    dependencies = []
+    for section in elf.iter_sections():
+        if not isinstance(section, NoteSection) or section.name != ".note.dlopen":
+            continue
+        for note in section.iter_notes():
+            if note["n_type"] != 0x407C0C0A or note["n_name"] != "FDO":
```
Comment (philiptaron) [inline pkgs/build-support/setup-hooks/auto-patchelf.py:68]: Both `0x407C0C0A` and `FDO` (which is short for Free Desktop Org) are documented at https://github.com/systemd/systemd/blob/main/docs/ELF_DLOPEN_METADATA.md.

**Review by philiptaron — APPROVED** [review]
It's too bad there is no existing place to land a unit test for these changes. 

I've run them successfully (along with the rest of the systemd 256.2 patch stack) on my actual machine, and can say that both in theory (such as I understand it) and in practice this change produces a working system.

**Comment by ElvishJerricco** [comment]
(updated with a change to the a comment's wording, as @layus suggested in the other PR)

**Review by philiptaron — APPROVED** [review]

File: pkgs/build-support/setup-hooks/auto-patchelf.py (hunk 1 above)
Comment (arianvp) [reply pkgs/build-support/setup-hooks/auto-patchelf.py:68]: There's already a comment at the top of the function explaining this.

**Review by nikstur — APPROVED** [review]

**Comment by ElvishJerricco** [comment]
Merged in #307068

**Pull request closed without merging**